envoy/network/listen_socket.h

```diff
@@ -87,5 +87,25 @@ class ConnectionSocket : public virtual Socket, public virtual ScopeTrackedObjec
 
 using ConnectionSocketPtr = std::unique_ptr<ConnectionSocket>;
 
+/**
+ * An interface through which a UDP socket, especially a QUIC socket, can
+ * postpone reading during hot restart until the parent instance is drained.
+ */
+class RegisterParentDrainedCallbackInterface {
+public:
+  /**
+   * @param address is the address of the listener. A std::string rather than a
+   *                string_view or reference to avoid a copy if the string was
+   *                constructed at the call-site.
+   * @param callback the function to call when the listener matching address is
+   *                 drained on the parent instance.
+   */
+  virtual void registerParentDrainedCallback(std::string address,
```

Comment:
Why is the address a string instead an address shared_ptr?

Reply:
Now a (const reference to) an address shared_ptr. It was an ill-considered simplification (it still ends up as a string in the end because we can't make a map keyed on address objects) - stringing it early reduces dependencies and makes it simpler to test, but also reduces how much it makes sense.